File: frontend/src/types/after-action-report.ts
export type AfterActionStatus = 'draft' | 'published' | 'archived';

export interface AfterActionTimelineEntry {
  timestamp: string;
  source: string;
  action: string;
  details: Record<string, unknown>;
}

export interface AfterActionKPI {
  response_time_minutes?: number;
  response_speed_score: number;
  dispatch_completion_rate: number;
  sitrep_coverage_rate: number;
  resource_efficiency_score: number;
  overall_score: number;
}

export interface AfterActionReport {
  id: string;
  report_code: string;
  emergency_event_id: string;
  title: string;
  summary: string;
  status: AfterActionStatus;
  timeline: AfterActionTimelineEntry[];
  kpi: AfterActionKPI;
  lessons_learned: string[];
  recommendations: string[];
  generated_by?: string;
  published_by?: string;
  generated_at: string;
  published_at?: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface AfterActionGenerateRequest {
  emergency_event_id: string;
  title?: string;
  summary_override?: string;
  lessons_learned?: string[];
  recommendations?: string[];
  force_regenerate?: boolean;
  metadata?: Record<string, unknown>;
}

export interface AfterActionUpdateRequest {
  title?: string;
  summary?: string;
  lessons_learned?: string[];
  recommendations?: string[];
  status?: AfterActionStatus;
  metadata?: Record<string, unknown>;
}

File: backend/app/domain/models/after_action_report.py
"""After-action report domain models."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from bson import ObjectId
from pydantic import BaseModel, Field, field_serializer, field_validator


class AfterActionStatus(str, Enum):
    """Lifecycle state of after-action reports."""

    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class AfterActionTimelineEntry(BaseModel):
    """One timeline entry in after-action report."""

    timestamp: datetime
    source: str
    action: str
    details: Dict[str, Any] = Field(default_factory=dict)


class AfterActionKPI(BaseModel):
    """Computed KPI values for post-incident evaluation."""

    response_time_minutes: Optional[float] = None
    response_speed_score: float = 0.0
    dispatch_completion_rate: float = 0.0
    sitrep_coverage_rate: float = 0.0
    resource_efficiency_score: float = 0.0
    overall_score: float = 0.0


class AfterActionReport(BaseModel):
    """After-action report aggregate."""

    id: Optional[str] = Field(
        validation_alias="_id",
        serialization_alias="id",
        default=None,
    )
    report_code: str
    emergency_event_id: str
    title: str
    summary: str
    status: AfterActionStatus = AfterActionStatus.DRAFT
    timeline: List[AfterActionTimelineEntry] = Field(default_factory=list)
    kpi: AfterActionKPI = Field(default_factory=AfterActionKPI)
    lessons_learned: List[str] = Field(default_factory=list)
    recommendations: List[str] = Field(default_factory=list)
    generated_by: Optional[str] = None
    published_by: Optional[str] = None
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    published_at: Optional[datetime] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    @field_validator("id", mode="before")
    @classmethod
    def validate_id(cls, value: Any):
        if isinstance(value, ObjectId):
            return str(value)
        return value

    @field_serializer("id")
    def serialize_id(self, value: Any, _info):
        if value is None:
            return None
        return str(value)

    class Config:
        populate_by_name = True
        arbitrary_types_allowed = True


class AfterActionGenerateRequest(BaseModel):
    """Payload for generating after-action report."""

    emergency_event_id: str
    title: Optional[str] = Field(default=None, min_length=5, max_length=240)
    summary_override: Optional[str] = None
    lessons_learned: List[str] = Field(default_factory=list)
    recommendations: List[str] = Field(default_factory=list)
    force_regenerate: bool = False
    metadata: Dict[str, Any] = Field(default_factory=dict)


class AfterActionReportUpdate(BaseModel):
    """Payload for updating draft after-action report."""

    title: Optional[str] = Field(default=None, min_length=5, max_length=240)
    summary: Optional[str] = None
    lessons_learned: Optional[List[str]] = None
    recommendations: Optional[List[str]] = None
    status: Optional[AfterActionStatus] = None
    metadata: Optional[Dict[str, Any]] = None
